### PR TITLE
Fix upgrade detection in Participating::new.

### LINKED ISSUE #2070: Make fast-sync work with `nctl`
We want to test the following using something like the nctl upgrade tests:

* start a network using 1.4 binaries
* upgrade the network to binaries using fast-sync
* join a fresh fast-sync node using a post-upgrade trusted hash
* ~join a fresh 1.4 node after the upgrade (should sync to upgrade point then upgrade)~

### PR DESCRIPTION
**Pull request by afck**
If after syncing we have no blocks, we expect to be at genesis, and immediately create block number 0.

If we do have blocks, and the highest one is the last one before the upgrade to the current version, we create the immediate switch block after the upgrade.

Closes #2070, #2370.

### CODE REVIEW

**Comment by afck** [comment]
Will retry CI once `dev` has been merged into `feat-fast-sync`.

File: node/src/reactor/participating.rs [hunk 1]
```diff
@@ -645,103 +704,53 @@ impl reactor::Reactor for Reactor {
...
                     trace!(%post_state_hash, ?execution_effect);
-                    initial_pre_state = ExecutionPreState::new(
+                    let initial_pre_state = ExecutionPreState::new(
                         0,
                         post_state_hash,
                         Default::default(),
                         Default::default(),
                     );
-                    finalized_block = FinalizedBlock::new(
+                    let finalized_block = FinalizedBlock::new(
```
Comment (Fraser999) [inline node/src/reactor/participating.rs]: Can't we use the new `create_immediate_switch_block()` function here?

Comment (afck) [reply node/src/reactor/participating.rs]: Of course! That's why I extracted the method! :facepalm:

File: node/src/reactor/participating.rs [hunk 2]
```diff
@@ -616,12 +614,73 @@ impl reactor::Reactor for Reactor {
         // If not, run genesis or upgrade and construct a switch block, and use that for the latest
         // block header.
         let latest_block_header = if let Some(latest_block_header) = maybe_latest_block_header {
-            latest_block_header
+            if latest_block_header.protocol_version() == chainspec.protocol_config.version {
+                latest_block_header
+            } else if latest_block_header.protocol_version() < chainspec.protocol_config.version
+                && latest_block_header.is_switch_block()
+                && ActivationPoint::EraId(latest_block_header.next_block_era_id())
+                    == chainspec.protocol_config.activation_point
```
Comment (rafal-ch) [inline node/src/reactor/participating.rs]: Please consider extracting this 3-condition `if` into a named function (`is_block_last_one_before_upgrade()`?) and unit-test it.

File: node/src/reactor/participating.rs (hunk 1 above)
Comment (afck) [reply node/src/reactor/participating.rs]: Done in https://github.com/casper-network/casper-node/pull/2416/commits/c222e53e6cbd8330e944b08ef4ebfe3e0714ec40.

**Review by rafal-ch — COMMENTED** [review]
LGTM,
However, I will check it against the nctl test prior to approval.

**Review by Fraser999 — APPROVED** [review]

File: node/src/reactor/participating.rs (hunk 2 above)
Comment (afck) [reply node/src/reactor/participating.rs]: Done in https://github.com/casper-network/casper-node/pull/2416/commits/e8775595f89eda0208ca5223820b487f6c27bff7.

**Review by fizyk20 — APPROVED** [review]
Nice one :+1:

**Comment by afck** [comment]
bors try

**Comment by TomVasile** [comment]
bors try

**Comment by afck** [comment]
bors merge

**Comment by afck** [comment]
bors merge

**Comment by afck** [comment]
bors merge

**Comment by afck** [comment]
bors merge

**Comment by afck** [comment]
bors merge

**Comment by afck** [comment]
bors merge

**Comment by afck** [comment]
Merging this: the other nctl upgrade test issues will be addressed separately.

**Pull request merged**